File: haskell-server/src/EagerWords/Server/Domain/Game.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE NamedFieldPuns #-}
{-# LANGUAGE DisambiguateRecordFields #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE FlexibleContexts #-}

module EagerWords.Server.Domain.Game (
    Game(..)
  , mkStartingGame
  , mkInitialGame
  , setPlayerTray
  , validatePlayAgainstGame
  , reflectPlayOnGame
  , doExchange
  , validateGameParams
  , gameAgeSeconds
  , toMiniState
  , summary
  , setBoard
  , setPlays
  , EagerWords.Server.Domain.Game.gameId
)
where

import Data.Int
import Data.Char (isAlphaNum)
import Data.List
import Data.Sequence (Seq, (><))
import qualified Data.Sequence as Seq
import Data.Time (UTCTime)
import Data.Time.Clock (diffUTCTime)
import Data.Time
import Data.Time.Clock.POSIX

import Control.Monad.IO.Class (MonadIO(..), liftIO)
import Control.Monad.Except (MonadError(..))

import qualified Bolour.Util.MiscUtil as Util
import Bolour.Util.TimeUtil (nowSecs)

import EagerWords.Common.Domain.GameParams as GameParams
import EagerWords.Common.Domain.InitPieces (InitPieces, InitPieces(InitPieces))
import qualified EagerWords.Common.Domain.InitPieces as InitPieces
import EagerWords.Common.Domain.PiecePoint (PiecePoint, PiecePoint(PiecePoint))
import qualified EagerWords.Common.Domain.PiecePoint as PiecePoint
import Bolour.Plane.Domain.Point (Point)
import qualified Bolour.Plane.Domain.Point as Point
import EagerWords.Common.Domain.Piece (Piece)
-- import EagerWords.Server.Domain.Player (Player)
import EagerWords.Server.Domain.Play (Play)
import qualified EagerWords.Server.Domain.Play as Play (mkWordPlay, mkSwapPlay)
import EagerWords.Common.Domain.PlayerType (PlayerType(..))
import qualified EagerWords.Common.Domain.PlayerType as PlayerType
import qualified EagerWords.Common.Domain.PlayerType as Player
import EagerWords.Common.Domain.PlayPiece (PlayPiece)
import qualified EagerWords.Common.Domain.PlayPiece as PlayPiece
import EagerWords.Common.Domain.GameMiniState (GameMiniState, GameMiniState(GameMiniState))
import EagerWords.Common.Domain.GameSummary (GameSummary, GameSummary(GameSummary))
import EagerWords.Common.Domain.StopInfo (StopInfo, StopInfo(StopInfo))

import EagerWords.Server.Domain.Board (Board)
import qualified EagerWords.Server.Domain.Board as Board
import EagerWords.Server.Domain.Tray (Tray, Tray(Tray))
import qualified EagerWords.Server.Domain.Tray as Tray
import EagerWords.Server.Domain.GameError
import EagerWords.Server.Domain.PieceProvider (PieceProvider)
import qualified EagerWords.Server.Domain.PieceProvider as PieceProvider
import qualified EagerWords.Server.Domain.Scorer as Scorer
import EagerWords.Server.Domain.GameBase (GameBase, GameBase(GameBase))
import qualified EagerWords.Server.Domain.GameBase as GameBase
import EagerWords.Common.Domain.GameSettings (GameSettings, GameSettings(GameSettings))
import qualified EagerWords.Common.Domain.GameSettings as GameSettings

data Game = Game {
    gameBase :: GameBase
  , board :: Board
  , trays :: [Tray.Tray] -- ^ Indexed by Player.playerTypeIndex UserPlayer or MachinePlayer. User 0, Machine 1.
  , playNumber :: Int
  , playTurn :: PlayerType
  , pieceProvider :: PieceProvider
  , scorePlay :: [PlayPiece] -> Int
  , lastPlayScore :: Int
  , scores :: [Int]
  , numSuccessivePasses :: Int
  , plays :: Seq Play
  , lastSecond :: Int64
}

noScore = 0
noScores = [0, 0]
initPlayNumber = 0

maxDimension = 120
maxTraySize = 26
maxSuccessivePasses = 10
noPasses = 0
noPlays :: Seq Play
noPlays = Seq.empty

gameId :: Game -> String
gameId Game {gameBase} = GameBase.gameId gameBase

mkStartingGame :: GameBase -> PieceProvider -> IO Game
mkStartingGame base pieceProvider = do
  let GameBase {gameParams, initPieces} = base
      GameParams {settings = GameSettings {dimension, trayCapacity}, pointValues} = gameParams
      InitPieces {userPieces, machinePieces, boardPiecePoints} = initPieces
      board = Board.mkBoardFromPiecePoints boardPiecePoints dimension
      scorer = Scorer.scorePlay $ Scorer.mkScorer pointValues
      initPlayerType = Player.User -- Won't be used. Arbitrarily set to user.
  lastSeconds <- nowSecs
  (userTray, pieceProvider') <- Tray.mkTray pieceProvider trayCapacity userPieces
  (machineTray, pieceProvider'') <- Tray.mkTray pieceProvider' trayCapacity machinePieces
  let initTrays = [userTray, machineTray]
      game = Game base board initTrays initPlayNumber initPlayerType pieceProvider''
          scorer noScore noScores noPasses noPlays lastSeconds
  return game

-- TODO. Complete show of game if needed.
instance Show Game where
  show game @ Game {board} = "Game {" ++ "board" ++ show board ++ "}"

passesMaxedOut :: Game -> Bool
passesMaxedOut Game { numSuccessivePasses } = numSuccessivePasses == maxSuccessivePasses

setBoard :: Game -> Board -> Game
setBoard game b = game {board = b}

setPlays :: Game -> Seq Play -> Game
setPlays game plays = game {plays = plays}

stopInfo :: Game -> StopInfo
stopInfo game @ Game { board, numSuccessivePasses } =
  StopInfo numSuccessivePasses (Board.isFilled board)

updatePieceGenerator :: Game -> PieceProvider -> Game

-- | Explicit record update for fieldGenerator.
--   Cannot do normal update: game {pieceProvider = nextGen} gets compiler error:
--   Record update for insufficiently polymorphic field: pieceProvider.
updatePieceGenerator game generator = game { pieceProvider = generator }

-- | Note. Validation of player name does not happen here.
mkInitialGame :: (MonadError GameError m, MonadIO m) =>
  GameParams -> InitPieces -> PieceProvider -> String -> m Game
mkInitialGame gameParams initPieces pieceProvider playerId = do
  base <- liftIO $ GameBase.mkInitialBase gameParams initPieces playerId
  liftIO $ mkStartingGame base pieceProvider

toMiniState :: Game -> GameMiniState
toMiniState game @ Game {board, pieceProvider, lastPlayScore, scores} =
  let
      gameEnded = passesMaxedOut game || Board.isFilled board
  in GameMiniState lastPlayScore scores gameEnded

gameAgeSeconds :: UTCTime -> Game -> Int64
gameAgeSeconds utcNow game @ Game {gameBase} =
  let GameBase {firstSecond} = gameBase
      nowSeconds = floor $ utcTimeToPOSIXSeconds utcNow
      ageSeconds = nowSeconds - firstSecond
  in ageSeconds

summary :: Game -> GameSummary
summary game @ Game {trays, scores} =
  let stopData = stopInfo game
  in GameSummary stopData

setPlayerTray :: Game -> PlayerType -> Tray -> Game
setPlayerTray (game @ Game {trays}) playerType tray =
  let whichTray = Player.playerTypeIndex playerType
      trays' = Util.setListElement trays whichTray tray
  in game {trays = trays'}

-- TODO. The following are only needed for player play.
-- Machine play already satisfies them.
-- TODO. The following are basic sanity checks. Not exhaustive yet. And some may be stubbed. Fix.
-- | Make sure the incoming play is consistent with the state of the game.
validatePlayAgainstGame :: MonadError GameError m => Game -> [PlayPiece] -> m [PlayPiece]
validatePlayAgainstGame game playPieces =
      checkContiguousPlay playPieces
  >>= checkPlayLineInBounds game
  >>= checkMoveDestinationsEmpty game
  >>= checkMoveTrayPieces game
  >>= checkPlayPositionsOccupied game
  >>= checkPlayBoardPieces game

-- TODO. Check connectivity. Either has to have an anchor or be a parallel play.
-- That is check in teh UI for now. So OK to defer.
checkContiguousPlay :: MonadError GameError m => [PlayPiece] -> m [PlayPiece]
checkContiguousPlay playPieces =
  let points = PlayPiece.point <$> playPieces
      contiguousBy f = Util.contiguous (f <$> points)
  in if contiguousBy Point.row || contiguousBy Point.col then
       return playPieces
     else throwError $ NonContiguousPlayError points

checkPlayLineInBounds :: MonadError GameError m => Game -> [PlayPiece] -> m [PlayPiece]
checkPlayLineInBounds _ [] = return []
checkPlayLineInBounds (Game { board }) playPieces = do
  Board.validatePoint board (PlayPiece.point (head playPieces))
  Board.validatePoint board (PlayPiece.point (last playPieces))
  return playPieces

-- Assume grid points have been checked to be in bounds.
checkMoveDestinationsEmpty :: MonadError GameError m => Game -> [PlayPiece] -> m [PlayPiece]
checkMoveDestinationsEmpty Game {board} playPieces =
  let gridPoints = PlayPiece.point <$> filter PlayPiece.moved playPieces
      maybeTaken = find (Board.pointIsNonEmpty board) gridPoints
  in case maybeTaken of
     Nothing -> return playPieces
     Just taken -> throwError $ OccupiedMoveDestinationError taken

-- TODO. Check cross words. Rename.
-- checkMoveDestinationsFreeCrossWise :: MonadError GameError m => Game -> [PlayPiece] -> m [PlayPiece]
-- checkMoveDestinationsFreeCrossWise game = return

-- TODO. Implement validation.
checkMoveTrayPieces :: MonadError GameError m => Game -> [PlayPiece] -> m [PlayPiece]
checkMoveTrayPieces game = return

-- | Check that the existing board pieces in the purported play are
--   in fact occupied already.
checkPlayPositionsOccupied :: MonadError GameError m => Game -> [PlayPiece] -> m [PlayPiece]
checkPlayPositionsOccupied Game {board} playPieces =
  let boardPlayPieces = filter (not . PlayPiece.moved) playPieces
      maybeFreePlayPiece = find (not . Board.pointHasValue board . PlayPiece.point) boardPlayPieces
  in case maybeFreePlayPiece of
     Nothing -> return playPieces
     Just missingPlayPiece -> throwError $ MissingBoardPlayPieceError $ PlayPiece.getPiecePoint missingPlayPiece

-- | Check that purported play pieces already on the board indicated on a play
--   do exist on the board.
checkPlayBoardPieces :: MonadError GameError m => Game -> [PlayPiece] -> m [PlayPiece]
checkPlayBoardPieces Game {board} playPieces =
  let boardPlayPieces = filter (not . PlayPiece.moved) playPieces
      piecePoints = PlayPiece.getPiecePoint <$> boardPlayPieces
      piecePointMatchesBoard PiecePoint {piece, point} =
        let maybePiece = Board.getPiece board point
        in case maybePiece of
           Nothing -> False
           Just boardPiece -> boardPiece == piece
      maybeMismatch = find (not . piecePointMatchesBoard) piecePoints
  in case maybeMismatch of
       Nothing -> return playPieces
       Just piecePoint -> throwError $ UnmatchedBoardPlayPieceError piecePoint

reflectPlayOnGame :: (MonadError GameError m, MonadIO m) =>
  Game -> PlayerType -> [PlayPiece] -> (Board -> (Board, [Point])) -> m (Game, [Piece], [Point])

reflectPlayOnGame game playerType playPieces deadPointFinder = do
  let Game {board, trays, playNumber, pieceProvider, numSuccessivePasses, scorePlay, scores, plays} = game
  _ <- if playerType == PlayerType.User then validatePlayAgainstGame game playPieces else return playPieces
  let movedPlayPieces = filter PlayPiece.moved playPieces
      movedPiecePoints = PlayPiece.getPiecePoint <$> movedPlayPieces
      b = Board.setPiecePoints board movedPiecePoints
      (b', deadPoints) = deadPointFinder b
      usedPieces = PiecePoint.piece <$> movedPiecePoints
      playerIndex = Player.playerTypeIndex playerType
  (newPieces, pieceProvider') <- liftIO $ PieceProvider.takePieces pieceProvider (length usedPieces)
  lastSecond <- liftIO nowSecs
  let game' = game {pieceProvider = pieceProvider'}
  let tray = trays !! playerIndex
      tray' = Tray.replacePieces tray usedPieces newPieces
      trays' = Util.setListElement trays playerIndex tray'
      playNumber' = playNumber + 1
      earlierScore = scores !! playerIndex
      thisScore = scorePlay playPieces
  let score' = earlierScore + thisScore
      scores' = Util.setListElement scores playerIndex score'
      play =  Play.mkWordPlay playNumber' playerType scores' playPieces newPieces deadPoints
      plays' = plays >< Seq.singleton play
      game'' = game' { board = b', trays = trays', playNumber = playNumber', lastPlayScore = thisScore, numSuccessivePasses = 0, 
      scores = scores', plays = plays', lastSecond = lastSecond}
  return (game'', newPieces, deadPoints)

doExchange :: (MonadError GameError m, MonadIO m) => Game -> PlayerType -> Int -> m (Game, Piece)
doExchange (game @ Game {board, trays, pieceProvider, playNumber, numSuccessivePasses, scores, plays}) playerType trayPos = do
  let whichTray = Player.playerTypeIndex playerType
      tray @ Tray {pieces} = trays !! whichTray
      piece = pieces !! trayPos -- TODO. Defensive check.
      succPasses = numSuccessivePasses + 1
      playNumber' = playNumber + 1
      game' = game { numSuccessivePasses = succPasses, lastPlayScore = 0 , playNumber = playNumber'}
  (piece', pieceProvider1) <- liftIO $ PieceProvider.swapOne pieceProvider piece
  lastSecond <- liftIO nowSecs
  -- TODO. Reduce constructions.
  let play =  Play.mkSwapPlay playNumber' playerType scores piece piece'
      plays' = plays >< Seq.singleton play
      game'' = game' { pieceProvider = pieceProvider1, plays = plays', lastSecond = lastSecond }
      tray' = Tray.replacePiece tray trayPos piece'
      game''' = setPlayerTray game'' playerType tray'
  return (game''', piece')

-- Either is a MonadError - but this generalizes the function.
-- In general rather than using a specific monad, use an mtl type class
-- that abstracts the specific functions provided by that monad.
-- Here rather than Either we use MonadError.
validateGameParams :: MonadError GameError m => GameParams -> m GameParams
validateGameParams params =
  let GameParams.GameParams {settings = GameSettings {dimension, trayCapacity}} = params
  in
    if dimension <= 0 || dimension > maxDimension then
      throwError $ InvalidDimensionError dimension
    else if trayCapacity <= 0 || trayCapacity > maxTraySize then
      throwError $ InvalidTrayCapacityError trayCapacity
    else return params

File: scala-server/app/com/bolour/eagerwords/server/domain/Game.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.server.domain

import com.bolour.eagerwords.common.domain.PlayerType._
import com.bolour.eagerwords.server.domain.GameExceptions.{InvalidCrosswordsException, MalformedPlayException}
import com.bolour.language.domain.WordDictionary
import com.bolour.eagerwords.common.domain.{PlayerType, _}
import GameState._
import com.bolour.eagerwords.server.service.GameServiceImpl.mkPieceProvider
import com.bolour.plane.domain.Point
import org.slf4j.LoggerFactory
import com.bolour.util.TimeUtil._

import scala.util.{Success, Try}

/**
  * The game.
  *
  * @param gameBase Data about the game that is conceptually immutable.
  * @param board The game's board.
  * @param trays The user and machine trays of pieces.
  * @param pieceProvider The factory for generating new pieces.
  * @param playNumber The ordinal number of the current play.
  * @param playTurn Whose play is it? User or machine.
  * @param lastPlayScore Score of the last play.
  * @param numSuccessivePasses Number of successive passes (if maxed out game ends).
  * @param scores The total scores of each player.
  * @param plays List of plays in the game so far.
  */
case class Game(
  gameBase: GameBase,
  board: Board,
  trays: List[Tray],
  pieceProvider: PieceProvider,
  playNumber: Int,
  playTurn: PlayerType,
  lastPlayScore: Int,
  numSuccessivePasses: Int,
  scores: List[Int],
  plays: Vector[Play],
  state: GameState = GameState.RUNNING,
  lastSecond: Long = nowSecs
) {
  import GameObj.MaxSuccessivePasses

  val logger = LoggerFactory.getLogger(this.getClass)
  val scorer = gameBase.scorer

  def isRunning = GameState.isRunning(state)
  def isCompleted = GameState.isCompleted(state)

  def passesMaxedOut: Boolean = numSuccessivePasses == MaxSuccessivePasses

  def isGuestGame: Boolean = gameBase.playerId.isEmpty

  def miniState: GameMiniState =
    GameMiniState(lastPlayScore, scores, passesMaxedOut || board.isFilled)

  def stopInfo: StopInfo = StopInfo(numSuccessivePasses, board.isFilled)

  // TODO. May fail. So return a Try for consistency.
  // For now nothing to do.
  def stop(): Game = {
    this
  }

  def summary(): GameSummary = GameSummary(stopInfo)

  def end(): Game = this.copy(
    lastSecond = nowSecs,
    state = GameState.ENDED
  )

  def suspend(): Game = this.copy(
    lastSecond = nowSecs,
    state = GameState.SUSPENDED
  )

  def resign(): Game = this.copy(
    lastSecond = nowSecs,
    state = GameState.RESIGNED
  )

  /**
    * Add the next play to the game, updating the board and returning
    * pieces to replace those used in the play.
    *
    * @param playerType User or machine.
    * @param playPieces Ordered list of pieces used in the play and their locations.
    *                   Includes all pieces that form the played word, and for each
    *                   an indication of whether they were moved in this play.
    * @param deadPointFinder Function to look for dead points after this play -
    *                        points that cannot possibly  be filled.
    * @return The updated game, a list of replacement pieces,
    *         and a list of dead points detected after the play is made.
    */
  def addWordPlay(playerType: PlayerType, playPieces: List[PlayPiece],
    deadPointFinder: Board => (Board, List[Point]) = GameObj.noDeads): Try[(Game, List[Piece], List[Point])] = {
    for {
      _ <- if (playerType == UserPlayer) validatePlay(playerType, playPieces) else Success(())
      movedPiecePoints = playPieces filter { _.moved } map { _.piecePoint }
      score = computePlayScore(playPieces)
      (playedGame, refills) <- addGoodWordPlay(playerType, movedPiecePoints, score)
      (newBoard, deadPoints) = deadPointFinder(playedGame.board)
      wordPlay = Play.mkWordPlay(playedGame.playNumber, playerType, playedGame.scores, playPieces, refills, deadPoints)
      playsPlus = plays :+ wordPlay
      newGame = playedGame.copy(board = newBoard, plays = playsPlus, lastSecond = nowSecs)
    } yield (newGame, refills, deadPoints)
  }

  def setDeadPoints(deadPoints: List[Point]): Game = {
    val newBoard = board.setBlackPoints(deadPoints)
    this.copy(board = newBoard, lastSecond = nowSecs)
  }

  def tray(playerType: PlayerType): Tray = trays(playerIndex(playerType))

  def computePlayScore(playPieces: List[PlayPiece]): Int = {
    scorer.scorePlay(playPieces)
  }

  /**
    * Add a validated play to the game and return replacements for the played pieces.
    */
  private def addGoodWordPlay(playerType: PlayerType, piecePoints: List[PiecePoint], score: Int): Try[(Game, List[Piece])] = {
    val newBoard = board.setPiecePoints(piecePoints)
    val usedPieces = piecePoints map { _.piece }
    val succPasses = if (score > 0) 0 else numSuccessivePasses + 1
    val ind = playerIndex(playerType)
    for {
      (nextProvider, newPieces) <- pieceProvider.takePieces(usedPieces.length)
      newTrays = trays.updated(ind, trays(ind).replacePieces(usedPieces, newPieces))
      newScores = scores.updated(ind, scores(ind) + score)
      nextType = nextPlayerType(playerType)
      newGame = Game(gameBase, newBoard, newTrays, nextProvider, playNumber + 1, nextType, score, succPasses, newScores, plays)
    } yield ((newGame, newPieces))
  }

  def addSwapPlay(piece: Piece, playerType: PlayerType): Try[(Game, Piece)] = {
    val succPasses = numSuccessivePasses + 1
    val trayIndex = PlayerType.playerIndex(playerType)
    val tray = trays(trayIndex)
    for {
      tray1 <- tray.removePiece(piece)
      (pieceProvide1, newPiece) <- pieceProvider.swapOne(piece)
      tray2 <- tray1.addPiece(newPiece)
      trays2 = trays.updated(playerIndex(playerType), tray2)
      newGame = this.copy(trays = trays2, pieceProvider = pieceProvide1, playNumber = playNumber + 1, numSuccessivePasses = succPasses, lastPlayScore = 0)
      swapPlay = Play.mkSwapPlay(newGame.playNumber, playerType, newGame.scores, piece, newPiece)
      playsPlus = plays :+ swapPlay
      newGamePlus = newGame.copy(plays = playsPlus, lastSecond = nowSecs)
    } yield (newGamePlus, newPiece)
  }

  /**
    * Check that cross words generated by this play are valid words in the dictionary.
    */
  def checkCrossWords(playPieces: List[PlayPiece], theDictionary: WordDictionary): Try[Unit] = {
    val theBoard = board
    val theTray = trays(playerIndex(UserPlayer))
    Try {
      val playStrip = board.stripOfPlay(playPieces)
      val stripMatcher = new StripMatcher {
        override def dictionary = theDictionary
        override def board = theBoard
        override def tray = theTray
      }
      val word = PlayPieceObj.playPiecesToWord(playPieces)
      val crosswords = stripMatcher.crossings(playStrip, word)
      val invalidCrosswords = crosswords.filter { w => !(theDictionary hasWord w) }
      if (!invalidCrosswords.isEmpty)
        throw new InvalidCrosswordsException(gameBase.gameParams.languageCode, invalidCrosswords)
    }
  }

  private def validatePlay(playerType: PlayerType, playPieces: List[PlayPiece]): Try[Unit] = {
    if (playerType == MachinePlayer)
      return Success(())

    for {
      _ <- checkPlayLineInBounds(playPieces)
      _ <- checkPlayAnchored(playPieces)
      _ <- checkContiguousPlay(playPieces)
      _ <- checkMoveDestinationsEmpty(playPieces)
      _ <- checkUnmovedPlayPositions(playPieces)
      _ <- checkMoveTrayPieces(playPieces)
    } yield ()
  }

  // TODO. Implement all play validation checks.
  // Most of the validations are also done in the associated web ui.
  // To allow unrelated client code to use the API, however, these checks
  // need to be implemented.

  def inBounds(coordinate: Int): Boolean =
    coordinate >= 0 && coordinate < gameBase.gameParams.dimension

  def inBounds(point: Point): Boolean = {
    inBounds(point.row) && inBounds(point.col)
  }
  /**
    * Check that play locations are within the board's boundaries.
    */
  private def checkPlayLineInBounds(playPieces: List[PlayPiece]): Try[Unit] = Try {
    val points = playPieces.map {_.point}
    if (!points.forall(inBounds(_)))
      throw MalformedPlayException("play points out of bounds")
  }

  // private def checkFirstPlayCentered(playPieces: List[PlayPiece]): Try[Unit] = Success(())

  /**
    * Check that the play is anchored - except for the first play.
    * TODO. Implement validation. Done in UI as well. So OK to defer.
    * Note. Anchored means the play includes at least one existing board piece,
    * or that the play is parallel. However, parallel plays will likely be removed.
    */
  private def checkPlayAnchored(playPieces: List[PlayPiece]): Try[Unit] = Success(())

  /**
    * Check that play for for a contiguous strip on the board.
    */
  private def checkContiguousPlay(playPieces: List[PlayPiece]): Try[Unit] = Try {
    val points = playPieces.map {_.point}
    if (points.length != points.distinct.length)
      throw MalformedPlayException("duplicate play points")

    val rows = points.map { _.row }
    val cols = points.map { _.col }

    // TODO. Move to util.
    def same[A](values: List[A]): Boolean =
      if (values.length < 2) true else values.drop(1).forall(_ == values.head)
    if (!same(rows) && !same(cols))
      throw MalformedPlayException("non-contiguous play points")
  }

  /**
    * Check that the destinations of the moved play pieces are empty.
    * TODO. Implement validation. Done in UI as well. So OK to defer.
    */
  private def checkMoveDestinationsEmpty(playPieces: List[PlayPiece]): Try[Unit] = Success(())

  /**
    * Check that the play pieces that are not moved and are supposed to exist on the board
    * are in fact there and contain the pieces indicated in the play.
    * TODO. Implement validation. Done in UI as well. So OK to defer.
    */
  private def checkUnmovedPlayPositions(playPieces: List[PlayPiece]): Try[Unit] = Success(())

  /**
    * Check that the moved pieces in this play are in fact user tray pieces.
    */
  private def checkMoveTrayPieces(playPieces: List[PlayPiece]): Try[Unit] = Success(())

  def sanityCheck: Try[Unit] = Try {
    val boardPieces = board.piecePoints.map { _.piece }
    val userTrayPieces = trays(playerIndex(UserPlayer)).pieces
    val machineTrayPieces = trays(playerIndex(MachinePlayer)).pieces

    val boardPieceIds = boardPieces.map { _.id }
    val userPieceIds = userTrayPieces.map { _.id }
    val machinePieceIds = machineTrayPieces.map { _.id }

    val allIds = boardPieceIds ++ userPieceIds ++ machinePieceIds
    val dups = allIds diff allIds.distinct
    if (dups.nonEmpty) {
      val message = s"duplicate ids in game state: ${dups}"
      logger.error(message)
      throw new IllegalStateException(message)
    }
  }
}

object GameObj {

  def MaxSuccessivePasses = 10

  def mkGame(gameBase: GameBase): Try[Game] = {

    val initPieces = gameBase.initPieces
    val board = Board(gameBase.gameParams.dimension, initPieces.boardPiecePoints)

    val pieceProvider: PieceProvider = mkPieceProvider(gameBase.gameParams.pieceProviderType)
    val lastPlayScore = 0

    for {
      (userTray, pieceGen1) <- mkTray(gameBase.gameParams.trayCapacity, initPieces.userPieces, pieceProvider)
      (machineTray, pieceGen2) <- mkTray(gameBase.gameParams.trayCapacity, initPieces.machinePieces, pieceGen1)
    }
      yield Game(gameBase, board, List(userTray, machineTray), pieceGen2, 0, UserPlayer, lastPlayScore, 0, List(0, 0), Vector.empty)
  }

  def mkTray(capacity: Int, initPieces: List[Piece], pieceGen: PieceProvider): Try[(Tray, PieceProvider)] = {
    if (initPieces.length >= capacity)
      return Success((Tray(capacity, initPieces.take(capacity).toVector), pieceGen))

    for {
      (nextGen, restPieces) <- pieceGen.takePieces(capacity - initPieces.length)
      pieces = initPieces ++ restPieces
    } yield (Tray(capacity, pieces.toVector), nextGen)
  }

  def noDeads(board: Board): (Board, List[Point]) = (board, Nil)

}
